crates/kind-query/src/metadata.rs

```diff
@@ -0,0 +1,90 @@
+use std::cell::RefCell;
+use std::ops::{Deref, DerefMut};
+use std::rc::Rc;
+use std::sync::Arc;
+
+use intmap::Entry;
+use refl::{refl, Id};
+
+use kind_syntax::concrete;
+use kind_syntax::core;
+
+use crate::build::{Compiler, Telemetry};
+use crate::loader::FileLoader;
+
+#[derive(Clone, Debug, Hash)]
+pub struct Source(pub Arc<String>);
+
+#[derive(Default)]
+pub struct Metadata(pub Vec<Storage>);
+
+#[derive(Clone)]
+pub enum Storage<A = Box<()>> {
```

Comment:
You need to add a doc string explaining that it's similar to a GADT because It's confusing.